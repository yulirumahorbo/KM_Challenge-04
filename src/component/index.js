import { View, StyleSheet, ActivityIndicator} from 'react-native'
import React from 'react'
import { moderateScale } from 'react-native-size-matters'

export default function Loading() {
  return (
    <View style={styles.loadingContainer}>
      <ActivityIndicator size={moderateScale(60)}/>
    </View>
  )
}

const styles = StyleSheet.create({
    loadingContainer: {
      flex: 1,
      justifyContent: 'center',
      alignItems: 'center',
    },
})