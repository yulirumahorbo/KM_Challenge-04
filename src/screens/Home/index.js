import {
  View,
  Text,
  TouchableOpacity,
  FlatList,
  Image,
  StyleSheet,  ActivityIndicator
} from 'react-native';
import React, {useEffect } from 'react';
import { moderateScale } from 'react-native-size-matters';
import { useDispatch, useSelector } from 'react-redux';
import {getBook, getBookDetail} from './redux/action';
import {setRefresh } from '../../reduxGlobal/action'

export default function Home() {
  const dispatch = useDispatch();
  const {bookData = []} = useSelector(state => state.home)
  const {loading, refresh} = useSelector(state => state.global)

  const getBooks = () => {
    dispatch(getBook())
  }

  const getBookDetails = item => {
    dispatch(getBookDetail(item.id))
  }

  useEffect(() => {
    getBooks();
  }, []);

  const listBook = ({ item }) => (
    <View style={{
      marginBottom: moderateScale(16),
      marginRight: moderateScale(16),
      height: moderateScale(200),
      width: moderateScale(100),
    }}
    >
    <TouchableOpacity onPress={()=>getBookDetails(item)}>
      <Image
        source={{ uri: `${item.cover_image}` }}
        style={{
          width: moderateScale(100),
          height: moderateScale(150),
          borderRadius: moderateScale(6),
          resizeMode: 'contain',
        }}
      />
        <Text style={[styles.regularSubText, {fontWeight: '600'}]} numberOfLines={1}>{item.title}</Text>
        <Text style={[styles.regularSubText, {fontSize: moderateScale(15),}]}>{item.author}</Text>
    </TouchableOpacity>

    </View>
  );

  if (loading) {
    return (
      <View flex={1} justifyContent="center" alignItems="center">
        <ActivityIndicator size={moderateScale(50)} />
      </View>
    );
  }
  const onRefresh = () => {
    dispatch(setRefresh(true))
    getBooks();
  }
  return (
    <View style={{ flex: 1, backgroundColor: 'white', padding: moderateScale(16)}}>
      <Text style={styles.regularSubText}>Good Morning, Veric!</Text>
      <View>
      <Text style={styles.regularText}>Recommended</Text>
        <FlatList
          style={{ marginTop: moderateScale(16) }}
          data={bookData}
          keyExtractor={(item, index) => index}
          renderItem={listBook}
          horizontal
          showsHorizontalScrollIndicator={false}
        />
      </View>
      <View>
      <Text style={styles.regularText}>Popular Book</Text>
        <FlatList
          refreshing={refresh}
          onRefresh={() => onRefresh()}
          numColumns={3}
          style={{ marginTop: moderateScale(16)}}
          data={bookData}
          keyExtractor={(item, index) => index}
          renderItem={listBook}
          showsVerticalScrollIndicator={false}
        />
      </View>

    </View>
  );
}

const styles = StyleSheet.create({
  regularText:{
    fontWeight: '500',
    fontSize: moderateScale(18),
    color: '#000000',
  },
  regularSubText:{
    fontWeight: '400',
    fontSize: moderateScale(16),
    color: '#000000',
  },
});
