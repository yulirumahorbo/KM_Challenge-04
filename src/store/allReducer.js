import {combineReducers} from 'redux';
import Global from '../reduxGlobal/reducer';
import HomeReducer from '../screens/Home/redux/reducer';
import LoginReducer from '../screens/Login/redux/reducer'

export const allReducer = combineReducers({
    login: LoginReducer,
    home: HomeReducer,
    global: Global,
})