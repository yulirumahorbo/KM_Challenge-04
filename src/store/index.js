import { createStore, applyMiddleware, combineReducers } from 'redux';
import logger from 'redux-logger';
import thunk from 'redux-thunk';
import { persistStore, persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import Global from '../reduxGlobal/reducer';
import HomeReducer from '../screens/Home/redux/reducer';
import LoginReducer from '../screens/Login/redux/reducer';

const allReducer = combineReducers({
  login: LoginReducer,
  home: HomeReducer,
  global: Global,
});

const config = {
  key: 'abcdefghijklmn',
  storage,
  timeout: null,
  whitelist: ['Login'],
};

const persistedReducer = persistReducer(config, allReducer);

const allMiddlewares = applyMiddleware(logger, thunk);

export const store = createStore(persistedReducer, {}, allMiddlewares);

export const persistedStore = persistStore(store);
