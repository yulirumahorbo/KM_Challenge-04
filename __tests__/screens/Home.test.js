import React from 'react';
import { create } from 'react-test-renderer';
import ContainerTesting from '../../src/helpers/reduxTesting';
import Home from '../../src/screens/Home';
import { store } from '../../src/store';

const Token = store.getState().login.tokenData;
const booksData = store.getState().home.bookData;

jest.useFakeTimers('legacy');

describe('Snapshot', () => {
  test('Home Snapshot', async () => {
    const component = create(ContainerTesting(<Home />));
    expect(component).toMatchSnapshot();
  });
});

describe('Validation Data Type', () => {
  test('BooksData', () => {
    expect(typeof booksData).toBe('object');
  });

  test('Token', () => {
    expect(typeof Token).toBe('string');
  });
});
