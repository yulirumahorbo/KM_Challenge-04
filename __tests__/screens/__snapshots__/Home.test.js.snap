// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Snapshot 1`] = `null`;

exports[`Home Snapshot should render 1`] = `null`;

exports[`Home should render 1`] = `null`;

exports[`Name of the group 1 should render 1`] = `null`;

exports[`Snapshot 1 1`] = `null`;

exports[`Snapshot Home Snapshot 1`] = `null`;
